R/private.R

```diff
@@ -0,0 +1,187 @@
+}
+
+
+#compares a reference vector of objects to a reference vector. If different it will return the difference.
+.isVectorContained = function(vectorToCompare = c(), vectorReference = c()){
+  x = vector(length = length(vectorToCompare))
+  for (i in 1:length(vectorToCompare)) {
+      x[i] = vectorToCompare[i] %in% vectorReference
+      if (length(which(vectorToCompare[i] %in% vectorReference)) == 0) {
+        vectorReference } else {
+          vectorReference = vectorReference[-match(vectorToCompare[i], vectorReference)]
+        }
+  }
+  if (all(x == T)) {
+    TRUE
+  } else {
+    return(vectorToCompare[!x])
+  }
+}
+
+# recursively flattens tree based structure.
+.flattenTree <- function(node, accumulated) {
+  if (is.null(node$children)) {
+    accumulated$name <- c(accumulated$name, node$name);
+    accumulated$size <- c(accumulated$size, node$size);
+    return(accumulated)
+  } else {
+    for (child in node$children) {
+      accumulated <- flattenTree(child, accumulated)
+    }
+    return(accumulated)
+  }
+}
+
+# converts time in integer/milliseconds to date-time with timezone.
+# assumption is that the system timezone = time zone of the local server running webApi.
+.millisecondsToDate <- function(milliseconds) {
+  sec <- milliseconds/1000
+  as.POSIXct(sec, origin = "1970-01-01", tz = Sys.timezone())
+}
+
+# checks if url conforms with expected structure for base url
+.checkWebApiBaseUrl <- function(baseUrl){
+  require(rex)
+  
+  valid_chars <- rex(except_some_of(".", "/", " ", "-"))
+  
+  baseUrlRegEx <- rex(
+    start,
+    
+    # protocol identifier (optional) + //
+    group(list("http", maybe("s")), "://"),
+    
+    # user:pass authentication (optional)
+    maybe(non_spaces,
+          maybe(":", zero_or_more(non_space)),
+          "@"),
+    
+    #host name
+    group(zero_or_more(valid_chars, zero_or_more("-")), one_or_more(valid_chars)),
+    
+    #domain name
+    zero_or_more(".", zero_or_more(valid_chars, zero_or_more("-")), one_or_more(valid_chars)),
+    
+    #TLD identifier
+    group(".", valid_chars %>% at_least(2)),
+    
+    # server port number (required)
+    zero_or_more(":", digit %>% between(2, 5)),
+    
+    # resource path (optional)
+    maybe("/", non_space %>% zero_or_more()),
+    
+    end
+  )
+  success <- as.logical(grepl(baseUrlRegEx, baseUrl))
+  if (!success) {
+    stop("Base URL not valid, should be like http://server.org:80/WebAPI")
+  }
+}
+
+# formats string/name
+.formatName <- function(name) {
+  gsub("_", " ", gsub("\\[(.*?)\\]_", "", gsub(" ", "_", name)))
+}
+
+# get valid source keys
+.getValidSourceKeys <- function(baseUrl,sourceKeys){
+  .checkWebApiBaseUrl(baseUrl)
+  require(dplyr)
+  .getCdmSources(baseUrl) %$%
+    parsed %>%
+    dplyr::filter(toupper(sourceKey) %in% toupper(sourceKeys)) %>%
+    dplyr::select(sourceKey) %>%
+    pull()
+}
+
+
+# get CDM sources
+.getCdmSources <- function(baseUrl) {
```

Comment:
I'd prefer we have 1 getCdmSources function rather than one public and now this private one. Seems like we could just merge the logic into the former.

Reply:
the general thought is that the public function needs to be deprecated and replaced with private function - because this function does not need to be end-user facing.

Reply:
the two functions are actually different. the public function getCdmSources returns cdmDatabaseSchema,  vocabDatabaseSchema, resultsDatabaseSchema - because these assigned in the function.

```
    cdmDatabaseSchema <- NA
    vocabDatabaseSchema <- NA
    resultsDatabaseSchema <- NA

```

The new private function, gets all the configuration information. It uses a transpose logic to create column names. column names are derived from field 'daimonType' in daimon table. The daimon types are CDM, Vocabulary, Results 

```
  daimons <- result$parsed$daimons %>%
    dplyr::bind_rows() %>%
    tidyr::spread(key = daimonType, value = tableQualifier) %>%
    dplyr::select(-sourceDaimonId) %>%
    base::replace(is.na(.), "") %>%
    dplyr::group_by(sourceId) %>%
    dplyr::summarise_all(max) %>%
    dplyr::mutate_all(na_if,"")
```
I recommend we retain this new function and the old one. We can think of a new name for new function.

Reply:
How about getSourceAndDaimonConfiguration ?